appinventor/appengine/src/com/google/appinventor/client/explorer/commands/RenameFileCommand.java

```diff
@@ -0,0 +1,165 @@
+import com.google.appinventor.client.OdeAsyncCallback;
+import com.google.appinventor.client.widgets.LabeledTextBox;
+import com.google.appinventor.client.youngandroid.TextValidators;
+import com.google.appinventor.shared.rpc.project.ProjectNode;
+import com.google.gwt.event.dom.client.*;
+import com.google.gwt.user.client.Command;
+import com.google.gwt.user.client.DeferredCommand;
+import com.google.gwt.user.client.ui.*;
+
+import static com.google.appinventor.client.Ode.MESSAGES;
+
+/**
+ * Command for renaming files.
+ */
+public class RenameFileCommand extends ChainableCommand {
+
+  @Override
+  protected boolean willCallExecuteNextCommand() {
+    return true;
+  }
+
+  @Override
+  protected void execute(ProjectNode node) {
+    new RenameDialog(node).center();
+  }
+
+  /**
+   * This class defines the dialog box for renaming an asset.
+   */
+  private class RenameDialog extends DialogBox {
+    // UI elements
+    private final LabeledTextBox newNameTextBox;
+    private ProjectNode node;
+
+    RenameDialog(ProjectNode node) {
+      super(false, true);
+      this.node = node;
+
+      setStylePrimaryName("ode-DialogBox");
+      setText(MESSAGES.renameTitle());
+      VerticalPanel contentPanel = new VerticalPanel();
+
+      LabeledTextBox oldNameTextBox = new LabeledTextBox(MESSAGES.oldNameLabel());
+      oldNameTextBox.setText(node.getName());
+      oldNameTextBox.setEnabled(false);
+      contentPanel.add(oldNameTextBox);
+
+      newNameTextBox = new LabeledTextBox(MESSAGES.newNameLabel());
+      newNameTextBox.getTextBox().addKeyUpHandler(new KeyUpHandler() {
+        @Override
+        public void onKeyUp(KeyUpEvent event) {
+          int keyCode = event.getNativeKeyCode();
+          if (keyCode == KeyCodes.KEY_ENTER) {
+            handleOkClick();
+          } else if (keyCode == KeyCodes.KEY_ESCAPE) {
+            hide();
+          }
+        }
+      });
+      contentPanel.add(newNameTextBox);
+
+      Button cancelButton = new Button(MESSAGES.cancelButton());
+      cancelButton.addClickHandler(new ClickHandler() {
+        @Override
+        public void onClick(ClickEvent event) {
+          hide();
+        }
+      });
+      Button okButton = new Button(MESSAGES.okButton());
+      okButton.addClickHandler(new ClickHandler() {
+        @Override
+        public void onClick(ClickEvent event) {
+          handleOkClick();
+        }
+      });
+      HorizontalPanel buttonPanel = new HorizontalPanel();
+      buttonPanel.add(cancelButton);
+      buttonPanel.add(okButton);
+      buttonPanel.setSize("100%", "24px");
+      contentPanel.add(buttonPanel);
+      contentPanel.setSize("320px", "100%");
+
+      add(contentPanel);
+    }
+
+    private void handleOkClick() {
+      final String newName = newNameTextBox.getText();
+      final String newFileId = node.getParent().getFileId() + "/" + newName;
+      // No change in name
+      if (newName == this.node.getName()) {
```

Comment:
We should also have a check to make sure that the new filename doesn't match an existing file.